### DISCUSSION TITLE
Add a Write-Only / Parameter Attribute for the DscProperty attribute

### DISCUSSION THREAD

**Discussion by michaeltlombardi** [PowerShell/PSDesiredStateConfiguration#76] [Ideas]
> As an engineer responsible for using and integrating with DSC Resources, I want to be able to learn from the metadata about a DSC Resource's property if it can/should be _passed to_ the resource but not expected to be _returned from_ the resource.

There is, currently, no way in DSC to distinguish between values that can be passed to the Resource _but not retrieved from it_ and those that can. The most common example of this type of property is a credential, but others typically exist to be used when a resource is created, to confirm boundaries, or to enforce set-time behavior.

The problem with not being able to distinguish between parameters (writable, not readable) and properties (writable and readable) surfaces particularly with integration tools. The only things those tools know about a DSC Resource is what the DSC Resource reports and, without a way to report that a value should not be expected to be read back, they have no way of knowing if something went silently wrong or a null value is correct behavior.

I propose that a new attribute be added to the list of existing attributes (**Key**, **Mandatory**, and **NotConfigurable**) for the `[DscProperty()]` attribute used to declare a class property as a DSC Resource property, called **Parameter**, **NotReturnable**, **WriteOnly**, or something similar (in descending preference order). This attribute should be used to declare that the property can be _passed_ to a resource but never _read back_ from it. It should be usable in conjunction with the existing **Mandatory** attribute.

While there is some nuance in whether or not there should be distinction at this level for whether the object can/should be passed to `Get` as well as `Set` and `Test` (e.g. a **Force** switch may not make sense for all three but a credential object might), that's worth considering more broadly and independent of the need for this information in the API surface of DSC Resources.

### Install-Time Parameters

Examples of parameter properties used in creation are install switches and the location of source files when installing a package for the first time; there's no way to read these values back from an extant resource.

The [`SqlServerDsc/SqlRSSetup`][dsc-SqlRSSetup] DSC Resource has numerous install switches - **IAcceptLicenseTerms**, **SourcePath**, **SupressRestart**, **SourceCredential**, **ProductKey**, **ForceRestart**, **EditionUpgrade**, **VersionUpgrade**, **Edition**, **LogPath**, and **SetupProcessTimeout**. These are _runtime_ options that can't be read back from an installed instance but the API surface of a DSC Resource has no way to describe them as such - they're all listed as writable properties.

The [Notes section of the implementation][dsc-SqlRSSetup-snip] even calls out that **ProductKey** and **Edition** can't be returned from the system:

```powershell
<#
    .NOTES
        The following properties are always returning $null because it's currently
        unknown how to return that information.
          - ProductKey
          - Edition
        The following properties always return $null on purpose. This could be
        changed in the future.
          - Action
          - SourceCredential
          - ForceRestart
          - EditionUpgrade
          - VersionUpgrade
          - LogPath
#>
```

### Boundary Declaring Parameters

An example of parameter properties used as boundaries are version limiters - while a DSC Resource can report the _current_ version of a DSC Resource, the minimum and maximum version boundaries exist as configuration options but not in the managed resource itself.

The [`PowerShellGet/PSModule`][dsc-PSModule] DSC Resource (in v2) uses the **RequiredVersion**, **MaximumVersion**, and **MinimumVersion** DSC Properties (with the `Write` attribute) to determine the desired version range for a module, but [the `Get` implementation][dsc-PSModule-snip] simply passes these values from the caller to the output because they can't be retrieved from the system:

```powershell
    $returnValue = @{
        Ensure             = 'Absent'
        Name               = $Name
        Repository         = $Repository
        Description        = $null
        Guid               = $null
        ModuleBase         = $null
        ModuleType         = $null
        Author             = $null
        InstalledVersion   = $null
        RequiredVersion    = $RequiredVersion
        MinimumVersion     = $MinimumVersion
        MaximumVersion     = $MaximumVersion
        Force              = $Force
        AllowClobber       = $AllowClobber
        SkipPublisherCheck = $SkipPublisherCheck
        InstallationPolicy = $null
    }
```

### Behavior Enforcing Parameters

An example of enforcing set-time behavior is the use of a **Force** switch parameter (for non-class-based Resources) or boolean property (for class-based Resources).

The [`SqlServerDsc/SqlAGDatabase`][dsc-SqlAGDatabase] DSC Resource uses the **Force** parameter for `Set` operations to "ensure the specified database(s) are the only databases that are a member of the specified Availability Group." In [the implementation][dsc-SqlAGDatabase-snip], the return object for `Get` sets the **Force** property to `$false` and does not update it from the system (because it _can't_):

```powershell
    $currentConfiguration = @{
        DatabaseName          = @()
        ServerName            = $ServerName
        InstanceName          = $InstanceName
        AvailabilityGroupName = ''
        BackupPath            = ''
        Ensure                = ''
        Force                 = $false
        MatchDatabaseOwner    = $false
        ReplaceExisting       = $false
        IsActiveNode          = $false
    }
```

## Integration Example

I ran into this particularly when writing Puppet.Dsc and ended up having to maintain a [utility function][puppet-dsc-test-parameter] to check if a property was defined as a credential or its name matched a hard-coded list of properties which were never returned in a `Get` request, i.e.

```powershell
  $KnownParameters = @(
    'Force'
    'JoinOU'
    'KeyUsage'
    'OID'
    'Purge'
    'Validate'
  )

  $Property.Name -in $KnownParameters -or $Property.ReferenceClassName -match 'Credential'
```

[puppet-dsc-test-parameter]: https://github.com/puppetlabs/Puppet.Dsc/blob/b90caabaff45676e807736f6d04fc2d2c0630741/src/internal/functions/Test-DscResourcePropertyParameterStatus.ps1
[dsc-SqlAGDatabase]: https://github.com/dsccommunity/SqlServerDsc/wiki/SqlAGDatabase#parameters
[dsc-SqlAGDatabase-snip]: https://github.com/dsccommunity/SqlServerDsc/blob/a9850f2a0d145c15a2a19d23bf20ff670fce17fa/source/DSCResources/DSC_SqlAGDatabase/DSC_SqlAGDatabase.psm1#L60-L72
[dsc-PSModule]: https://github.com/PowerShell/PowerShellGetv2/blob/master/DSC/DscResources/MSFT_PSModule/MSFT_PSModule.schema.mof
[dsc-PSModule-snip]: https://github.com/PowerShell/PowerShellGetv2/blob/ddaee667fba6d04cfd0534d6abb544e1f7a7ce48/DSC/DscResources/MSFT_PSModule/MSFT_PSModule.psm1#L106-L123
[dsc-SqlRSSetup]: https://github.com/dsccommunity/SqlServerDsc/wiki/SqlRSSetup
[dsc-SqlRSSetup-snip]: https://github.com/dsccommunity/SqlServerDsc/blob/a9850f2a0d145c15a2a19d23bf20ff670fce17fa/source/DSCResources/DSC_SqlRSSetup/DSC_SqlRSSetup.psm1#L26-L39

**Comment by kilasuit** [comment]
I like this idea and do think it would be a good addition 🙂
